File: js-port/funcscript-js/src/funcs/math/division-function.js
const { BaseFunction, CallType } = require('../../core/function-base');
const { ensureTyped, typeOf, valueOf, makeValue } = require('../../core/value');
const { FSDataType } = require('../../core/fstypes');
const { makeError, FsError } = require('../helpers');

function isNumericType(type) {
  return type === FSDataType.Integer || type === FSDataType.Float || type === FSDataType.BigInteger;
}

class DivisionFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = '/';
    this.callType = CallType.Infix;
    this.precidence = 50;
  }

  evaluate(provider, parameters) {
    if (parameters.count === 0) {
      return makeError(FsError.ERROR_PARAMETER_COUNT_MISMATCH, `${this.symbol}: at least one operand required`);
    }

    let current = ensureTyped(parameters.getParameter(provider, 0));
    let mode = typeOf(current);

    if (!isNumericType(mode)) {
      mode = FSDataType.Integer;
      current = makeValue(FSDataType.Integer, 1);
    }

    let intTotal = mode === FSDataType.Integer ? valueOf(current) : 1;
    let longTotal = mode === FSDataType.BigInteger ? valueOf(current) : 1n;
    let doubleTotal = mode === FSDataType.Float ? valueOf(current) : 1.0;

    const zeroError = () => makeError(FsError.ERROR_TYPE_INVALID_PARAMETER, `${this.symbol}: division by zero`);

    const promoteIntToLong = () => {
      if (mode !== FSDataType.Integer) {
        return;
      }
      mode = FSDataType.BigInteger;
      longTotal = BigInt(intTotal);
    };

    const promoteIntToDouble = () => {
      if (mode === FSDataType.Float) {
        return;
      }
      mode = FSDataType.Float;
      doubleTotal = Number(intTotal);
    };

    const promoteLongToDouble = () => {
      if (mode === FSDataType.Float) {
        return;
      }
      mode = FSDataType.Float;
      doubleTotal = Number(longTotal);
    };

    const divideDouble = (divisor) => {
      if (divisor === 0) {
        return zeroError();
      }
      doubleTotal /= divisor;
      return null;
    };

    const divideInt = (divisor) => {
      if (divisor === 0) {
        return zeroError();
      }
      if (intTotal % divisor === 0) {
        intTotal = Math.trunc(intTotal / divisor);
        return null;
      }
      promoteIntToDouble();
      return divideDouble(divisor);
    };

    const divideLong = (divisor) => {
      if (divisor === 0n) {
        return zeroError();
      }
      if (longTotal % divisor === 0n) {
        longTotal = longTotal / divisor;
        return null;
      }
      promoteLongToDouble();
      return divideDouble(Number(divisor));
    };

    for (let i = 1; i < parameters.count; i += 1) {
      const operand = ensureTyped(parameters.getParameter(provider, i));
      const operandType = typeOf(operand);

      if (!isNumericType(operandType)) {
        return makeError(FsError.ERROR_TYPE_MISMATCH, `${this.symbol}: numeric operand expected`);
      }

      let error = null;
      if (mode === FSDataType.Integer) {
        if (operandType === FSDataType.Integer) {
          error = divideInt(valueOf(operand));
        } else if (operandType === FSDataType.BigInteger) {
          promoteIntToLong();
          error = divideLong(valueOf(operand));
        } else {
          promoteIntToDouble();
          error = divideDouble(Number(valueOf(operand)));
        }
      } else if (mode === FSDataType.BigInteger) {
        if (operandType === FSDataType.Float) {
          promoteLongToDouble();
          error = divideDouble(Number(valueOf(operand)));
        } else {
          const divisor = operandType === FSDataType.BigInteger ? valueOf(operand) : BigInt(valueOf(operand));
          error = divideLong(divisor);
        }
      } else {
        error = divideDouble(Number(valueOf(operand)));
      }

      if (error) {
        return error;
      }
    }

    if (mode === FSDataType.Float) {
      return makeValue(FSDataType.Float, doubleTotal);
    }
    if (mode === FSDataType.BigInteger) {
      return makeValue(FSDataType.BigInteger, longTotal);
    }
    return makeValue(FSDataType.Integer, intTotal);
  }
}

module.exports = {
  DivisionFunction
};

File: FuncScript/Functions/Math/DivisionFunction.cs
using FuncScript.Core;
using FuncScript.Model;

namespace FuncScript.Functions.Math
{
    public class DivisionFunction : IFsFunction
    {
        public int MaxParsCount => -1;

        public CallType CallType => CallType.Infix;

        public string Symbol => "/";

        public int Precedence => 50;

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            var ret = EvaluateInteral(pars, (i) =>
            {
                var ret = pars.GetParameter(parent, i);
                return (true, ret);
            });
            return ret;
        }
        object EvaluateInteral(IParameterList pars,Func<int,(bool,object)> getPar)
        {
            bool isInt = false, isLong = false, isDouble = false;
            int intTotal = 1;
            long longTotal = 1;
            double doubleTotal = 1;
            int count = pars.Count;

            if (count > 0)
            {
                var p = getPar(0);
                if (!p.Item1)
                    return null;
                var d = p.Item2;

                if (d is int)
                {
                    isInt = true;
                    intTotal = (int)d;
                }
                else if (d is long)
                {
                    isLong = true;
                    longTotal = (long)d;
                }
                else if (d is double)
                {
                    isDouble = true;
                    doubleTotal = (double)d;
                }
                else
                {
                    isInt = true;
                    intTotal = 1;
                }
            }

            for (int i = 1; i < count; i++)
            {
                var p = getPar(i);
                if (!p.Item1)
                    return null;
                var d = p.Item2;

                if (isInt)
                {
                    if (d is int intDiv)
                    {
                        DivideInt(intDiv);
                        continue;
                    }

                    if (d is long longDiv)
                    {
                        PromoteIntToLong();
                        DivideLong(longDiv);
                        continue;
                    }

                    if (d is double doubleDiv)
                    {
                        PromoteIntToDouble();
                        doubleTotal /= doubleDiv;
                        continue;
                    }
                }

                if (isLong)
                {
                    if (d is int intDiv)
                    {
                        DivideLong(intDiv);
                        continue;
                    }

                    if (d is long longDiv)
                    {
                        DivideLong(longDiv);
                        continue;
                    }

                    if (d is double doubleDiv)
                    {
                        PromoteLongToDouble();
                        doubleTotal /= doubleDiv;
                        continue;
                    }
                }

                if (isDouble)
                {
                    if (d is int intDiv)
                    {
                        doubleTotal /= intDiv;
                    }
                    else if (d is long longDiv)
                    {
                        doubleTotal /= longDiv;
                    }
                    else if (d is double doubleDiv)
                    {
                        doubleTotal /= doubleDiv;
                    }
                }
            }

            if (isDouble)
                return doubleTotal;
            if (isLong)
                return longTotal;
            if (isInt)
                return intTotal;

            return null;

            void PromoteIntToLong()
            {
                if (!isInt)
                    return;
                isInt = false;
                isLong = true;
                longTotal = intTotal;
            }

            void PromoteIntToDouble()
            {
                if (isDouble)
                {
                    isInt = false;
                    return;
                }

                isInt = false;
                isDouble = true;
                doubleTotal = intTotal;
            }

            void PromoteLongToDouble()
            {
                if (isDouble)
                {
                    isLong = false;
                    return;
                }

                isLong = false;
                isDouble = true;
                doubleTotal = longTotal;
            }

            void DivideInt(int divisor)
            {
                if (intTotal % divisor == 0)
                {
                    intTotal /= divisor;
                }
                else
                {
                    PromoteIntToDouble();
                    doubleTotal /= divisor;
                }
            }

            void DivideLong(long divisor)
            {
                if (longTotal % divisor == 0)
                {
                    longTotal /= divisor;
                }
                else
                {
                    PromoteLongToDouble();
                    doubleTotal /= divisor;
                }
            }
        }

        public string ParName(int index)
        {
            return $"Op {index + 1}";
        }
    }
}
